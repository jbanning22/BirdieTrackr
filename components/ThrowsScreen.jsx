import {
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
  FlatList,
  SafeAreaView,
} from 'react-native';
import React, {useState, useEffect} from 'react';
import moment from 'moment';
import AsyncStorage from '@react-native-async-storage/async-storage';
import axios from 'axios';
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome';
import {faTrashCan} from '@fortawesome/free-solid-svg-icons/faTrashCan';
import {Alert} from 'react-native';

const ThrowsScreen = ({navigation}) => {
  const [throwData, setThrowData] = useState([]);

  const getThrows = async () => {
    const token = await AsyncStorage.getItem('token');
    const headers = {
      Authorization: `Bearer ${token}`,
    };
    try {
      const measuredThrows = await axios.get(
        'http://localhost:3000/measure-throws',
        {headers},
      );
      setThrowData(measuredThrows.data);
    } catch (error) {
      //   console.log(error);
      console.log('error getting throw details');
    }
  };

  const deleteThrow = async id => {
    const token = await AsyncStorage.getItem('token');
    const headers = {
      Authorization: `Bearer ${token}`,
    };
    Alert.alert(
      'Delete Throw',
      'Are you sure you want to delete this throw?',
      [
        {text: 'Cancel', style: 'cancel'},
        {
          text: 'Delete',
          onPress: async () => {
            try {
              const deleteThrow = await axios.delete(
                `http://localhost:3000/measure-throws/${id}`,
                {headers},
              );
              getThrows();
            } catch (error) {
              console.log(error);
            }
          },
        },
      ],
      {cancelable: false},
    );
  };

  const renderItem = ({item}) => {
    const date = moment(item.createdAt).format('MMMM Do, YYYY');
    return (
      <View style={styles.flatListParent}>
        <View style={styles.flatlistStyle}>
          <Text style={styles.renderCourseName}>{item.distance} ft</Text>
          <Text style={styles.renderHoleText}>{item.disc}</Text>
          <Text style={styles.throwTypeText}>{item.throwtype}</Text>
          <Text style={styles.renderText}>{date}</Text>
        </View>
        <TouchableOpacity onPress={() => deleteThrow(item.id)}>
          <FontAwesomeIcon
            icon={faTrashCan}
            color={'white'}
            size={12}
            style={{margin: 4}}
          />
        </TouchableOpacity>
      </View>
    );
  };
  useEffect(() => {
    getThrows();
  }, []);
  //   useEffect(() => {
  //     getThrows();
  //   }, []);
  //   const sortedThrows = throwData.sort((a, b) => b.distance - a.distance);
  return (
    <SafeAreaView style={styles.box1}>
      <Text style={styles.titleText}>Throws</Text>
      {throwData.length === 0 ? (
        <Text>You have not recorded any throws yet.</Text>
      ) : (
        <FlatList
          renderItem={renderItem}
          data={throwData}
          showsVerticalScrollIndicator={false}
        />
      )}
      <TouchableOpacity
        style={styles.measureThrowButton}
        onPress={() => navigation.navigate('ThrowsScreen2')}>
        <Text style={styles.buttonText}>Measure a Throw</Text>
      </TouchableOpacity>
    </SafeAreaView>
  );
};

export default ThrowsScreen;

const styles = StyleSheet.create({
  box1: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#DB6F52',
  },
  titleText: {
    fontSize: 36,
    alignSelf: 'center',
    fontWeight: '500',
    color: 'white',
    marginBottom: 20,
  },
  flatListParent: {
    flexDirection: 'row',
    backgroundColor: '#52BEDB',
    justifyContent: 'center',
    alignContent: 'center',
    borderRadius: 8,
    marginBottom: 10,
  },
  measureThrowButton: {
    width: 200,
    height: 50,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#52BEDB',
    marginTop: 10,
    borderRadius: 8,
    marginBottom: 20,
  },
  buttonText: {
    color: 'white',
    fontSize: 16,
    fontWeight: '500',
  },
  flatlistStyle: {
    width: 250,
    marginLeft: 15,
  },
  renderItemStyle: {
    flexDirection: 'column',
    margin: 10,
  },
  renderCourseName: {
    alignSelf: 'center',
    fontSize: 26,
    fontWeight: '600',
    color: 'white',
  },
  renderHoleText: {
    alignSelf: 'center',
    fontSize: 18,
    fontWeight: '500',
    color: 'white',
  },
  throwTypeText: {
    alignSelf: 'center',
    fontSize: 16,
    fontWeight: '400',
    color: 'white',
  },
  renderText: {
    alignSelf: 'center',
    fontSize: 12,
    fontWeight: '400',
    marginBottom: 5,
    color: 'white',
  },
});
