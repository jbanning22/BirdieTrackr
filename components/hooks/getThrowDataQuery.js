import axios from 'axios';
import {useQuery} from '@tanstack/react-query';
import AsyncStorage from '@react-native-async-storage/async-storage';

const getThrowsUrl =
  'http://ec2-54-87-189-240.compute-1.amazonaws.com:3000/measure-throws';

const getThrows = async () => {
  const token = await AsyncStorage.getItem('token');
  const headers = {
    Authorization: `Bearer ${token}`,
  };
  try {
    const response = await axios.get(getThrowsUrl, {headers});
    return response ? response.data : null;
  } catch (error) {
    throw error;
  }
};

export const useGetThrowDetails = () => {
  const {data, isLoading, isError, error} = useQuery(['throwData'], getThrows);
  return {data, isLoading, isError, error};
};
