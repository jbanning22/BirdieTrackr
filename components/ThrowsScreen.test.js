import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import ThrowsScreen from './ThrowsScreen.jsx';
import {QueryClient, QueryClientProvider} from '@tanstack/react-query';
import {useGetThrowDetails} from './hooks/getThrowDataQuery.js';

jest.mock('axios');
jest.mock('@fortawesome/react-native-fontawesome', () => ({
  FontAwesomeIcon: () => null,
}));
jest.mock('./hooks/getThrowDataQuery.js', () => ({
  useGetThrowDetails: jest.fn(),
}));

describe('Throws Screen', () => {
  let queryClient;
  beforeEach(() => {
    queryClient = new QueryClient();
    queryClient.setDefaultOptions({queries: {cacheTime: 0}});
    const mockThrowData = [
      {
        id: 5,
        createdAt: '2023-05-31T19:15:14.085Z',
        disc: 'Test',
        color: 'Test',
        throwtype: 'Test',
        distance: '281',
        userId: 1,
      },
    ];
    useGetThrowDetails.mockReturnValue({
      data: mockThrowData,
      isLoading: false,
      isError: false,
      error: null,
    });
  });
  afterEach(() => {
    queryClient.clear();
    jest.clearAllMocks();
  });
  it('should match the snapshot', () => {
    const {toJSON} = render(
      <QueryClientProvider client={queryClient}>
        <ThrowsScreen />
      </QueryClientProvider>,
    );

    expect(toJSON()).toMatchSnapshot();
  });
  it('should display screen title', () => {
    const {getByText} = render(
      <QueryClientProvider client={queryClient}>
        <ThrowsScreen />
      </QueryClientProvider>,
    );
    expect(getByText('Throws')).toBeTruthy();
  });
  it('should display measure throw button', async () => {
    const {getByText} = render(
      <QueryClientProvider client={queryClient}>
        <ThrowsScreen />
      </QueryClientProvider>,
    );
    expect(getByText('Measure a Throw')).toBeTruthy();
  });
  it('should navigate to create throw screen', async () => {
    const navigation = {
      navigate: jest.fn(),
    };
    const {getByText} = render(
      <QueryClientProvider client={queryClient}>
        <ThrowsScreen navigation={navigation} />
      </QueryClientProvider>,
    );
    fireEvent.press(getByText('Measure a Throw'));
    expect(navigation.navigate).toHaveBeenCalledWith('ThrowsScreen2');
  });
});
