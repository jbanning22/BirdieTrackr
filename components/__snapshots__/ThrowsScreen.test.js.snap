// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Throws Screen should match the snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/images/BasketBackground2.png",
        }
      }
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          {
            "height": "100%",
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <Text
      style={
        {
          "alignSelf": "center",
          "backgroundColor": "white",
          "color": "black",
          "fontFamily": "Satoshi-Medium",
          "fontSize": 36,
          "fontWeight": "500",
        }
      }
    >
      Throws
    </Text>
    <View
      style={
        {
          "alignContent": "center",
          "alignItems": "center",
          "height": 570,
          "justifyContent": "center",
          "padding": 20,
          "width": 750,
        }
      }
    >
      <RCTScrollView
        data={
          [
            {
              "color": "Test",
              "createdAt": "2023-05-31T19:15:14.085Z",
              "disc": "Test",
              "distance": "281",
              "id": 5,
              "throwtype": "Test",
              "userId": 1,
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        showsVerticalScrollIndicator={false}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                {
                  "alignContent": "flex-start",
                  "backgroundColor": "#2D6061",
                  "borderRadius": 8,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginBottom": 10,
                }
              }
            >
              <View
                style={
                  {
                    "margin": 10,
                    "width": 280,
                  }
                }
              >
                <Text
                  style={
                    {
                      "alignSelf": "center",
                      "color": "white",
                      "fontFamily": "Satoshi-Medium",
                      "fontSize": 26,
                      "fontWeight": "600",
                    }
                  }
                >
                  281
                   ft
                </Text>
                <Text
                  style={
                    {
                      "color": "white",
                      "fontFamily": "Satoshi-Medium",
                      "fontSize": 18,
                      "fontWeight": "500",
                    }
                  }
                >
                  Test
                </Text>
                <Text
                  style={
                    {
                      "color": "white",
                      "fontFamily": "Satoshi-Medium",
                      "fontSize": 16,
                      "fontWeight": "500",
                    }
                  }
                >
                  Test
                </Text>
                <Text
                  style={
                    {
                      "color": "white",
                      "fontFamily": "Satoshi-Medium",
                      "fontSize": 12,
                      "fontWeight": "500",
                      "marginBottom": 5,
                    }
                  }
                >
                  May 31st, 2023
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#2D6061",
          "borderRadius": 14,
          "height": 56,
          "justifyContent": "center",
          "marginBottom": 15,
          "opacity": 1,
          "width": 327,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "fontFamily": "Satoshi-Medium",
            "fontSize": 18,
            "fontWeight": "500",
          }
        }
      >
        Measure a Throw
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
